keps/sig-node/3610-env-injection/README.md

```diff
@@ -0,0 +1,920 @@
+
+5. Be able to specify system wide environment variables. We have multiple openshift instances (different environments and different datacenters). It would be great if that data could be made automatically available without having to deal with custom env sections in configs.
+
+6. `http_proxy`: We have a proxy in our datacenter and we need to set `http_proxy` for all containers. <https://github.com/kubernetes/kubernetes/issues/48180#issuecomment-356866173>
+
+7. `KUBERNETES_SERVICE_HOST`: Being able to set KUBERNETES_SERVICE_HOST cluster wide makes it possible to use an external hostname (with a publicly signed SSL certificate) without any issues. <https://github.com/kubernetes/kubernetes/issues/48180#issuecomment-375111483>
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- if the env is not set in container, inject env to all containers in namespace
+- if it is already set in container, override some env and not override for others
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- inject env to all containers in cluster
+- checking if configmap/secret in `valueFrom` is valid or not
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Add a new Admission Controller: GlobalEnv. Users can enable or disable it on the apiserver side.
+
+### Proposal 1: the initial proposal using annotations
+
+If enabled, the admission controller will add env according to namespace annotation to every pod's containers in the namespace wide.
+A namespace annotation named `global.env`, works for pods in the namespace.
+
+1. If a pod has already set the env, skip it. Overwriting is too aggressive.
+If users want to overwrite the env, `global.env/overwrite: true` should be set in namespace annotation.
+
+2. Env injection only for pod creation. Not for update(The behavior that a pod update may change its env, is odd).
+
+3. If the annotation is not set properly, skip it. A warning log or event will be helpful.
+This is mentioned in the risks as we have no valition for annotation by default.
+
+
+### Proposal 2: the second/final proposal using API Object
+
+If enabled, the admission controller will add env according to the API Object to every pod's containers in the namespace wide.
+We support both `defaultEnv` and `forceEnv` in the spec.
+
+- `defaultEnv` will be added to the container if the env is not set in the container.
+- `forceEnv` will override the env in the container.
```

Comment:
Should there be an option to suggest ignoring injection of this at the pod level ? Certain usecases like proxy, for example. 

If you have a pod needs proxy because it is reaching out to the cluster, but other pods don't need them because they are cluster internal. Injecting the proxy to all pods can cause problems if one misconfigures the entry such as `no_proxy` that can cause the cluster breakdown. So, would it be good to have a way to annotate a pod and say I am not interested in any of the env injection ?  Something `global.env/ignore: true`?

Reply:
This is another use case. Should I add it to the stories below?

For ValidatingAdmissionPolicy, they use binding mode to add a new binding with `matchResources:matchLabels`.
We may add a matchLabels field in GlobalEnv. By default, no matchLabels means all.

Reply:
> This is another use case. Should I add it to the stories below?

Would be good to have I think. Having a way to exclude injection can be equally important as having a way to inject something. 

> By default, no matchLabels means all.

Sure. This will also be inline with how rest of the things do the match. So definitely better than what I mentioned above via the annotation.